### PR TITLE
Prepare nodejs for v18

### PR DESCRIPTION
**Pull request by KozhinovAlexander**
Prepare nodjs package for version 18

---

- [x] I've read the [contributing guidelines](https://github.com/conan-io/conan-center-index/blob/master/CONTRIBUTING.md).
- [x] I've used a [recent](https://github.com/conan-io/conan/releases/latest) Conan client version close to the [currently deployed](https://github.com/conan-io/conan-center-index/blob/master/.c3i/config_v1.yml#L6).
- [x] I've tried at least one configuration locally with the [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

**Review by SSE4 — DISMISSED** [review]

File: recipes/nodejs/all/conanfile.py [hunk 1]
```diff
@@ -1,48 +1,82 @@
...
+    def source(self):
+        get(self, **self.conan_data["sources"][self.version][str(self.info.settings.os)][self._nodejs_arch], destination=self._source_subfolder, strip_root=True)
 
     def package(self):
-        self.copy(pattern="LICENSE", dst="licenses", src=self._source_subfolder)
-        self.copy(pattern="*", src=os.path.join(self._source_subfolder, "bin"), dst="bin")
-        self.copy(pattern="node.exe", src=self._source_subfolder, dst="bin")
-        self.copy(pattern="npm", src=self._source_subfolder, dst="bin")
-        self.copy(pattern="npx", src=self._source_subfolder, dst="bin")
+        if Version(__version__) < '2.0.0':
```
Comment (prince-chrismc) [inline recipes/nodejs/all/conanfile.py:66]: This is not required you can just use the new helper for 1.x

File: recipes/nodejs/all/conanfile.py [hunk 2]
```diff
@@ -1,48 +1,82 @@
 import os
-from conans import ConanFile, tools
-from conans.errors import ConanInvalidConfiguration
+import re
+from conans import __version__
```
Comment (prince-chrismc) [inline recipes/nodejs/all/conanfile.py:3]:
```suggestion
```

File: recipes/nodejs/all/conanfile.py [hunk 3]
```diff
@@ -1,48 +1,82 @@
...
+from six import StringIO
+
+required_conan_version = ">=1.59.0"
 
 
 class NodejsConan(ConanFile):
     name = "nodejs"
     description = "nodejs binaries for use in recipes"
-    topics = ("conan", "node", "nodejs")
+    topics = ("node", "nodejs")
```
Comment (prince-chrismc) [inline recipes/nodejs/all/conanfile.py:17]: repeating the name does not help the search :)

```suggestion
    topics = ("node", "javascript", "runtime")
```

File: recipes/nodejs/all/conanfile.py [hunk 4]
```diff
@@ -1,48 +1,82 @@
...
 
     @property
     def _source_subfolder(self):
         return os.path.join(self.source_folder, "source_subfolder")
 
     @property
     def _nodejs_arch(self):
-        if str(self.settings.os) == "Linux":
-            if str(self.settings.arch).startswith("armv7"):
+        if str(self.info.settings.os) == "Linux":
```
Comment (prince-chrismc) [inline recipes/nodejs/all/conanfile.py:31]: `self.info` should only be used in `package_info` this looks very bizarre 🤔

File: recipes/nodejs/all/conanfile.py [hunk 5]
```diff
@@ -1,48 +1,82 @@
...
+            if str(self.info.settings.arch).startswith("armv7"):
                 return "armv7"
-            if str(self.settings.arch).startswith("armv8") and "32" not in str(self.settings.arch):
+            if str(self.info.settings.arch).startswith("armv8") and "32" not in str(self.info.settings.arch):
                 return "armv8"
-        return str(self.settings.arch)
+        return str(self.info.settings.arch)
+
+    @property
+    def _glibc_version(self):
```
Comment (prince-chrismc) [inline recipes/nodejs/all/conanfile.py:39]: Why is this required?

File: recipes/nodejs/all/conanfile.py [hunk 6]
```diff
@@ -1,48 +1,82 @@
...
+        self.run(['ldd', '--version'], output=buff)
+        buff = re.search('GLIBC \d{1,3}.\d{1,3}', buff.getvalue())
+        return str(buff.group()[6:])
 
     def validate(self):
-        if (not (self.version in self.conan_data["sources"]) or
-            not (str(self.settings.os) in self.conan_data["sources"][self.version]) or
-            not (self._nodejs_arch in self.conan_data["sources"][self.version][str(self.settings.os)] ) ):
+        if not self.version in self.conan_data["sources"] or \
+           not str(self.info.settings.os) in self.conan_data["sources"][self.version] or \
```
Comment (prince-chrismc) [inline recipes/nodejs/all/conanfile.py:47]: info should not be add here 🤔

File: recipes/nodejs/all/conanfile.py [hunk 7]
```diff
@@ -1,48 +1,82 @@
...
+           not self._nodejs_arch in self.conan_data["sources"][self.version][str(self.info.settings.os)]:
             raise ConanInvalidConfiguration("Binaries for this combination of architecture/version/os not available")
 
+        if Version(self.version) >= "18.0.0":
+            if str(self.info.settings.compiler) == "gcc" and Version(self.info.settings.compiler.version) < "8.3":
+                raise ConanInvalidConfiguration("Binaries for this combination of architecture/version/os not available")
```
Comment (prince-chrismc) [inline recipes/nodejs/all/conanfile.py:53]: Why does precompiled nodejs binaries need a higher version of GCC? that seems out of place

File: recipes/nodejs/all/test_package/conanfile.py [hunk 8]
```diff
@@ -1,12 +1,13 @@
-import os
-from conans import ConanFile, tools
+from conan import ConanFile
+from conan.tools.build import cross_building
 
 
 class TestPackageConan(ConanFile):
 
     settings = "os", "arch"
+    tool_requires = "nodejs/[*]@"
```
Comment (prince-chrismc) [inline recipes/nodejs/all/test_package/conanfile.py:8]: This should be using `self_tested_reference` which you can find example of in the docs's template for pre compiled libraries

**Review by prince-chrismc — COMMENTED** [review]
Thanks for taking the time to submit this PR, there's a quiet a fe things i was not expecting so there are lots of questions if you dont mind answering that would be great

File: recipes/nodejs/all/conanfile.py (hunk 5 above)
Comment (KozhinovAlexander) [reply recipes/nodejs/all/conanfile.py:39]: Test will fail otherwise. See supported platforms here: https://github.com/nodejs/node/blob/main/BUILDING.md#platform-list It states glib>=2.17

File: recipes/nodejs/all/conanfile.py (hunk 6 above)
Comment (KozhinovAlexander) [reply recipes/nodejs/all/conanfile.py:47]: Without info linter complains about self.settings usage.

**Comment by KozhinovAlexander** [comment]
> Thanks for taking the time to submit this PR, there's a quiet a fe things i was not expecting so there are lots of questions if you dont mind answering that would be great

Thank you for your review. I've added some changes and comments. Some parts of your review I need to investigate later.

File: recipes/nodejs/all/test_package/conanfile.py (hunk 8 above)
Comment (KozhinovAlexander) [reply recipes/nodejs/all/test_package/conanfile.py:8]: You're right. It is removed now.

File: recipes/nodejs/all/conanfile.py (hunk 7 above)
Comment (KozhinovAlexander) [reply recipes/nodejs/all/conanfile.py]: You're right. It is removed now.

File: recipes/nodejs/all/test_package/conanfile.py (hunk 8 above)
Comment (KozhinovAlexander) [reply recipes/nodejs/all/test_package/conanfile.py]: Shall be resolved by: 6f4138b99ba39e6f7de52015104733b12186b012

File: recipes/nodejs/all/conanfile.py (hunk 1 above)
Comment (KozhinovAlexander) [reply recipes/nodejs/all/conanfile.py]: Shall be sloved by 2121d7acb9b0ef42d2d9305a112990479136328f

File: recipes/nodejs/all/conanfile.py (hunk 4 above)
Comment (KozhinovAlexander) [reply recipes/nodejs/all/conanfile.py:30]: See the reason here: https://github.com/conan-io/conan/issues/13506 Basically it is a bug in conan itself ;)

**Comment by KozhinovAlexander** [comment]
> The following fixes the build for Conan v2
> ```diff
> diff --git a/recipes/nodejs/all/conanfile.py b/recipes/nodejs/all/conanfile.py
> index 726d342df..37511ef45 100644
> --- a/recipes/nodejs/all/conanfile.py
> +++ b/recipes/nodejs/all/conanfile.py
> @@ -59,20 +59,12 @@ class NodejsConan(ConanFile):
>          get(self, **self.conan_data["sources"][self.version][str(self.info.settings.os)][self._nodejs_arch], destination=self._source_subfolder, strip_root=True)
>  
>      def package(self):
> -        if conan_version.major == "1":
> -            self.copy(pattern="LICENSE", dst="licenses", src=self._source_subfolder)
> -            self.copy(pattern="*", src=os.path.join(self._source_subfolder, "bin"), dst="bin")
> -            self.copy(pattern="*", src=os.path.join(self._source_subfolder, "lib"), dst="lib")
> -            self.copy(pattern="node.exe", src=self._source_subfolder, dst="bin")
> -            self.copy(pattern="npm", src=self._source_subfolder, dst="bin")
> -            self.copy(pattern="npx", src=self._source_subfolder, dst="bin")
> -        else:
> -            copy(self, pattern="LICENSE", dst="licenses", src=self._source_subfolder)
> -            copy(self, pattern="*", src=os.path.join(self._source_subfolder, "bin"), dst="bin")
> -            copy(self, pattern="*", src=os.path.join(self._source_subfolder, "lib"), dst="lib")
> -            copy(self, pattern="node.exe", src=self._source_subfolder, dst="bin")
> -            copy(self, pattern="npm", src=self._source_subfolder, dst="bin")
> -            copy(self, pattern="npx", src=self._source_subfolder, dst="bin")
> +        copy(self, pattern="LICENSE", dst=os.path.join(self.package_folder, "licenses"), src=self._source_subfolder)
> +        copy(self, pattern="*", dst=os.path.join(self.package_folder, "bin"), src=os.path.join(self._source_subfolder, "bin"))
> +        copy(self, pattern="*", dst=os.path.join(self.package_folder, "lib"), src=os.path.join(self._source_subfolder, "lib"))
> +        copy(self, pattern="node.exe", dst=os.path.join(self.package_folder, "bin"), src=self._source_subfolder)
> +        copy(self, pattern="npm", dst=os.path.join(self.package_folder, "bin"), src=self._source_subfolder)
> +        copy(self, pattern="npx", dst=os.path.join(self.package_folder, "bin"), src=self._source_subfolder)
>  
>      def package_info(self):
>          self.cpp_info.includedirs = []
> diff --git a/recipes/nodejs/all/test_package/conanfile.py b/recipes/nodejs/all/test_package/conanfile.py
> index 74e435003..26c1600e2 100644
> --- a/recipes/nodejs/all/test_package/conanfile.py
> +++ b/recipes/nodejs/all/test_package/conanfile.py
> @@ -7,8 +7,8 @@ class TestPackageConan(ConanFile):
>      settings = "os", "arch"
>      test_type = "explicit"
>  
> -    def requirements(self):
> -        self.requires(self.tested_reference_str)
> +    def build_requirements(self):
> +        self.tool_requires(self.tested_reference_str)
>  
>      def test(self):
>          if can_run(self):
> ```

Thank you for looking ob that. 
Please do code-suggestion next time. Then it is easier to take the code over.
Unfortunafely, if you will try your changes v1 and v2 both will complain about failed test (e.g. node --version). Looking at the problem more carefully you'll find, that copy(self, ... does not copy anything in package folder. At least it was the reason for my approach. 
BTW.: There ist already copy(self, .... for v2 in package() in the second part of If statement.

File: recipes/nodejs/all/conanfile.py [hunk 9]
```diff
@@ -1,48 +1,78 @@
...
+            self.copy(pattern="node.exe", src=self._source_subfolder, dst="bin")
+            self.copy(pattern="npm", src=self._source_subfolder, dst="bin")
+            self.copy(pattern="npx", src=self._source_subfolder, dst="bin")
+        else:
+            copy(self, pattern="LICENSE", dst="licenses", src=self._source_subfolder)
+            copy(self, pattern="*", src=os.path.join(self._source_subfolder, "bin"), dst="bin")
+            copy(self, pattern="*", src=os.path.join(self._source_subfolder, "lib"), dst="lib")
+            copy(self, pattern="node.exe", src=self._source_subfolder, dst="bin")
+            copy(self, pattern="npm", src=self._source_subfolder, dst="bin")
+            copy(self, pattern="npx", src=self._source_subfolder, dst="bin")
```
Comment (gjasny) [inline recipes/nodejs/all/conanfile.py:75]: The problem I see is not the `self.copy` vs. `copy(self` but the `dst` parameter. The actual folder must be prefixed with `self.package_folder`. See the [packaging example](https://github.com/conan-io/conan-center-index/blob/413775941483fbe74b7005c0864c6cc6fa755f3f/docs/package_templates/header_only/all/conanfile.py#L88)

```suggestion
        copy(self, pattern="LICENSE", dst=os.path.join(self.package_folder, "licenses"), src=self._source_subfolder)
        copy(self, pattern="*", dst=os.path.join(self.package_folder, "bin"), src=os.path.join(self._source_subfolder, "bin"))
        copy(self, pattern="*", dst=os.path.join(self.package_folder, "lib"), src=os.path.join(self._source_subfolder, "lib"))
        copy(self, pattern="node.exe", dst=os.path.join(self.package_folder, "bin"), src=self._source_subfolder)
        copy(self, pattern="npm", dst=os.path.join(self.package_folder, "bin"), src=self._source_subfolder)
        copy(self, pattern="npx", dst=os.path.join(self.package_folder, "bin"), src=self._source_subfolder)

```

Comment (KozhinovAlexander) [reply recipes/nodejs/all/conanfile.py:75]: > The problem I see is not the `self.copy` vs. `copy(self` but the `dst` parameter. The destination folder must be prefixed with `self.package_folder`. See the packaging example: https://github.com/conan-io/conan-center-index/blob/413775941483fbe74b7005c0864c6cc6fa755f3f/docs/package_templates/header_only/all/conanfile.py#L88
> 
> 
> 
> ```suggestion
> 
>         copy(self, pattern="LICENSE", dst=os.path.join(self.package_folder, "licenses"), src=self._source_subfolder)
> 
>         copy(self, pattern="*", dst=os.path.join(self.package_folder, "bin"), src=os.path.join(self._source_subfolder, "bin"))
> 
>         copy(self, pattern="*", dst=os.path.join(self.package_folder, "lib"), src=os.path.join(self._source_subfolder, "lib"))
> 
>         copy(self, pattern="node.exe", dst=os.path.join(self.package_folder, "bin"), src=self._source_subfolder)
> 
>         copy(self, pattern="npm", dst=os.path.join(self.package_folder, "bin"), src=self._source_subfolder)
> 
>         copy(self, pattern="npx", dst=os.path.join(self.package_folder, "bin"), src=self._source_subfolder)
> 
> 
> 
> ```

Thank you very much. Let's give it a try :)

**Comment by KozhinovAlexander** [comment]
@gjasny Thanks. It works.

File: recipes/nodejs/all/conanfile.py [hunk 10]
```diff
@@ -1,48 +1,70 @@
...
                 return "armv8"
-        return str(self.settings.arch)
+        return str(self.info.settings.arch)
+
+    @property
+    def _glibc_version(self):
+        buff = StringIO()
+        self.run(['ldd', '--version'], output=buff)
+        buff = re.search('GLIBC \d{1,3}.\d{1,3}', buff.getvalue())
+        return str(buff.group()[6:])
```
Comment (gjasny) [inline recipes/nodejs/all/conanfile.py:42]:
```suggestion
        buff = re.search(r'GLIBC (\d{1,3}.\d{1,3})', buff.getvalue())
        return str(buff.group(1))
```

File: recipes/nodejs/all/conanfile.py [hunk 11]
```diff
@@ -1,48 +1,70 @@
 import os
-from conans import ConanFile, tools
-from conans.errors import ConanInvalidConfiguration
+import re
+from conan import ConanFile, conan_version
```
Comment (gjasny) [inline recipes/nodejs/all/conanfile.py:3]: According to the linter `conan_version` is now unused
```suggestion
from conan import ConanFile
```

**Review by gjasny — APPROVED** [review]

**Comment by KozhinovAlexander** [comment]
@SSE4  @prince-chrismc Please take a look.

**Review by SSE4 — DISMISSED** [review]

File: recipes/nodejs/all/conanfile.py (hunk 4 above)
Comment (prince-chrismc) [reply recipes/nodejs/all/conanfile.py:30]: No no, it's usually a sign of doing something funky that probably has a different solution.

There are checks that need to happen in both generate and package_id

File: recipes/nodejs/all/conanfile.py (hunk 6 above)
Comment (prince-chrismc) [reply recipes/nodejs/all/conanfile.py:46]: No no, that issue was specific to using it in `package` for validate the recommendation is to access `self.settings` directly.

File: recipes/nodejs/all/conanfile.py [hunk 12]
```diff
@@ -1,48 +1,70 @@
...
+            if str(self.info.settings.os) == "Linux":
+                if Version(self._glibc_version) < '2.27':
+                    raise ConanInvalidConfiguration("Binaries for this combination of architecture/version/os not available")
+
     def build(self):
-        tools.get(**self.conan_data["sources"][self.version][str(self.settings.os)][self._nodejs_arch], destination=self._source_subfolder, strip_root=True)
+        pass
+
+    def source(self):
+        get(self, **self.conan_data["sources"][self.version][str(self.info.settings.os)][self._nodejs_arch], destination=self._source_subfolder, strip_root=True)
```
Comment (prince-chrismc) [inline recipes/nodejs/all/conanfile.py:59]: This should be using `self.settings` here as well

**Review by prince-chrismc — COMMENTED** [review]
There's some incorrect usage of `self.info` here that needs to be corrected, thats for package compatibility which I do not see any here 🤔

**Comment by KozhinovAlexander** [comment]
> There's some incorrect usage of `self.info` here that needs to be corrected, thats for package compatibility which I do not see any here 🤔 

Please point it out. If linter Not complains - it is difficult to find, what so you mean.

**Comment by KozhinovAlexander** [comment]
> There's some incorrect usage of `self.info` here that needs to be corrected, thats for package compatibility which I do not see any here thinking

Basically changed with following commit: adf292c6922f82a735570f915e643b6d5b217fe6

But with conan V2 I am getting following error on my local machine:

```bash
======== Installing packages ========
nodejs/16.3.0: Calling source() in /home/<user_name>/.conan2/p/nodej0214a1acb894f/s
ERROR: nodejs/16.3.0: Error in source() method, line 59
        get(self, **self.conan_data["sources"][self.version][str(self.settings.os)][self._nodejs_arch], destination=self._source_subfolder, strip_root=True)
        ConanException: 'self.settings' access in 'source()' method is forbidden
``

**Comment by KozhinovAlexander** [comment]
> There's some incorrect usage of `self.info` here that needs to be corrected, thats for package compatibility which I do not see any here thinking

@prince-chrismc It shall be fixed in 349d964876996cb99367b7f63107a14d3c92af21 and adf292c6922f82a735570f915e643b6d5b217fe6

**Comment by prince-chrismc** [comment]
Oh wow. the recipe was just bad before? pre compiled binaries should be downloaded in the build folder not the source method

Check the cmake recipe for an example https://github.com/conan-io/conan-center-index/blob/2dc2a4fa215e47d613a38b911cd781cde16e404f/recipes/cmake/binary/conanfile.py#L30

File: recipes/nodejs/all/conanfile.py [hunk 13]
```diff
@@ -1,19 +1,28 @@
...
     homepage = "https://nodejs.org"
     license = "MIT"
-    settings = "os", "arch"
+    settings = "os", "arch", "compiler"
     no_copy_source = True
+    short_paths = True
+
+    def init(self):
+        self.__os = None
+        self.__arch = None
```
Comment (prince-chrismc) [inline recipes/nodejs/all/conanfile.py:25]:
```suggestion
    short_paths = True
```

File: recipes/nodejs/all/conanfile.py [hunk 14]
```diff
@@ -28,21 +37,39 @@ def _nodejs_arch(self):
...
+           not self._nodejs_arch in self.conan_data["sources"][self.version][str(self.settings.os)]:
             raise ConanInvalidConfiguration("Binaries for this combination of architecture/version/os not available")
 
+        if Version(self.version) >= "18.0.0":
+            if str(self.settings.os) == "Linux":
+                if Version(self._glibc_version) < '2.27':
+                    raise ConanInvalidConfiguration("Binaries for this combination of architecture/version/os not available")
+
+        self.__os = self.settings.os
+        self.__arch = self._nodejs_arch
```
Comment (prince-chrismc) [inline recipes/nodejs/all/conanfile.py:57]:
```suggestion
                    raise ConanInvalidConfiguration("Binaries for this combination of architecture/version/os not available")
```

File: recipes/nodejs/all/conanfile.py [hunk 15]
```diff
@@ -28,21 +37,39 @@ def _nodejs_arch(self):
...
+
     def build(self):
-        tools.get(**self.conan_data["sources"][self.version][str(self.settings.os)][self._nodejs_arch], destination=self._source_subfolder, strip_root=True)
+        pass
+
+    def source(self):
```
Comment (prince-chrismc) [inline recipes/nodejs/all/conanfile.py:62]:
```suggestion
    def build(self):
```

File: recipes/nodejs/all/conanfile.py [hunk 16]
```diff
@@ -28,21 +37,39 @@ def _nodejs_arch(self):
...
+        get(self, **self.conan_data["sources"][self.version][str(self.__os)][self.__arch],
+            destination=self._source_subfolder, strip_root=True)
```
Comment (prince-chrismc) [inline recipes/nodejs/all/conanfile.py:64]:
```suggestion
        get(self, **self.conan_data["sources"][self.version][str(self.settings.os)][self.settings.arch],
            destination=self._source_subfolder, strip_root=True)
```

**Comment by KozhinovAlexander** [comment]
> Oh wow. the recipe was just bad before? pre compiled binaries should be downloaded in the build folder not the source method
> 
> Check the cmake recipe for an example
> 
> https://github.com/conan-io/conan-center-index/blob/2dc2a4fa215e47d613a38b911cd781cde16e404f/recipes/cmake/binary/conanfile.py#L30

Sometimes I am wondering the restrictions you tell me :) I've never seen thouse in conan docs. And I am using conan v1 already 1,5 years in my company :D

**Comment by gjasny** [comment]
> Sometimes I am wondering the restrictions you tell me :) I've never seen thouse in conan docs. And I am using conan v1 already 1,5 years in my company :D

Acyually there is a package template named [prebuild_tool_package](https://github.com/conan-io/conan-center-index/blob/master/docs/package_templates/prebuilt_tool_package/), which does it the same way: https://github.com/conan-io/conan-center-index/blob/eccffc6d90e68e5de79ba5673c465b74a42dea6a/docs/package_templates/prebuilt_tool_package/all/conanfile.py#L36-L46

I always try to follow one of the templates.

File: recipes/nodejs/all/test_package/conanfile.py [hunk 17]
```diff
@@ -1,12 +1,16 @@
...
+
+    def build_requirements(self):
+        self.tool_requires(self.tested_reference_str)
 
     def test(self):
-        if not tools.cross_building(self):
+        if can_run(self):
             self.output.info("Node version:")
-            self.run("node --version", run_environment=True)
+            self.run("node --version")
```
Comment (gjasny) [inline recipes/nodejs/all/test_package/conanfile.py:16]:
```suggestion
            self.run("node --version")
            self.run("npm --version")
            self.run("npx --version")
```

File: recipes/nodejs/all/test_v1_package/conanfile.py [hunk 18]
```diff
@@ -0,0 +1,16 @@
...
+    settings = "os", "arch"
+    test_type = "explicit"
+
+    def requirements(self):
+        self.requires(self.tested_reference_str)
+
+    def test(self):
+        if not cross_building(self):
+            self.output.info("Node version:")
+            self.run("node --version")
```
Comment (gjasny) [inline recipes/nodejs/all/test_v1_package/conanfile.py:16]:
```suggestion
            self.run("node --version")
            self.run("npm --version")
            self.run("npx --version")
```

**Review by gjasny — APPROVED** [review]

**Comment by gjasny** [comment]
And thank you for all your work and patience!  🐝

**Comment by prince-chrismc** [comment]
> Sometimes I am wondering the restrictions you tell me :) I've never seen thouse in conan docs. And I am using conan v1 already 1,5 years in my company :D

The biggest sin of 1.x was letting you do anything even thought it was not recommended or error prone.
In CCI we have 100s of PRs a week and to double checking every corner case is just not possible...
I personally just ask author to change it to the way I know it's going to work so we do not have awkward work arounds... just because you can does not mean you should.

Reviewing PRs is not my job but I actually enjoy it so I make a few hours a week to do them. 

As for the docs, they are never complete of finished so there's a done of things missing. Completely features are just undocumented still so as much as they are a source of truth they are sometimes missing things.

On the flip side CCI's community has a difference of opinion when it comes to writing recipes, not all of the conventions here are the "recommendation" from Conan itself... for example how we apply patches... it's easier to review and maintain the CCI way but not necessarily required.

But it's super hard to explain and looks and review for corner cases.

Lastly you have all the enterprise users who build packages from CCI, they prefer constancy so they can more easily patch and fix large amount of recipes at once. We do not support this but the more uniform a recipe gets the easier they re to wrk with.

**Comment by KozhinovAlexander** [comment]
> > Sometimes I am wondering the restrictions you tell me :) I've never seen thouse in conan docs. And I am using conan v1 already 1,5 years in my company :D
> 
> The biggest sin of 1.x was letting you do anything even thought it was not recommended or error prone.
> In CCI we have 100s of PRs a week and to double checking every corner case is just not possible...
> I personally just ask author to change it to the way I know it's going to work so we do not have awkward work arounds... just because you can does not mean you should.
> 
> Reviewing PRs is not my job but I actually enjoy it so I make a few hours a week to do them. 
> 
> As for the docs, they are never complete of finished so there's a done of things missing. Completely features are just undocumented still so as much as they are a source of truth they are sometimes missing things.
> 
> On the flip side CCI's community has a difference of opinion when it comes to writing recipes, not all of the conventions here are the "recommendation" from Conan itself... for example how we apply patches... it's easier to review and maintain the CCI way but not necessarily required.
> 
> But it's super hard to explain and looks and review for corner cases.
> 
> Lastly you have all the enterprise users who build packages from CCI, they prefer constancy so they can more easily patch and fix large amount of recipes at once. We do not support this but the more uniform a recipe gets the easier they re to wrk with.

Thank you very much for detailed explanation and taking time for review. Basically I am okay with it and happy to do the things right way. Some stuff was just very new to me, I've never seen in that context. But now, I can say, that a lot of developers writing recipes ...hmmh... not completely right :D

**Comment by prince-chrismc** [comment]
❤️  Just trying to fill in the context since this project has changed over the last almost 4 years :)

**Review by prince-chrismc — APPROVED** [review]

**Comment by KozhinovAlexander** [comment]
@gjasny @SSE4 would you please approve?

**Review by gjasny — APPROVED** [review]

**Review by SSE4 — APPROVED** [review]
